Add Governance doc (#9167)
wpilibsuite/allwpilib: Governance.md

# **WPILib Governance Structure**

> ## **Background**

Reviewer: Some more feedback on the Markdown export; we don't need every heading doubly-bolded right?

## **Steering Council**

### **Purpose**

The Steering Council will be the ultimate authoritative decision-making body in the WPILib project for the purpose of:

* Setting broad project direction including long-term roadmaps, approving proposed Major changes each year, etc.
* Managing membership of the Core Developers
* Approving changes to core project guidelines such as Code of Conduct, Governance, Mission, and Privacy Policy
* Resolving any disputes

The goal is for the Steering Council to formally exercise these powers as little as is feasible, preferring instead to build consensus within the Core Developers and community whenever possible. When exercising these powers formally, the Steering Council should aim to do so by consensus, but may establish direction via majority vote if consensus cannot be reached.

### **Composition**

The Steering Council is composed of 3 members representing the diverse groups with substantial interest in the development of WPILib:

> FIRST – 1 individual representing FIRSTs interest in WPILib through its use as the core programming library in the *FIRST* Robotics Competition and *FIRST* Tech Challenge (FTC) programs.

Reviewer 2: ```suggestion
FIRST – 1 individual representing the interests of FIRST in WPILib through its use as the core programming library in the *FIRST* Robotics Competition and *FIRST* Tech Challenge (FTC) programs.
```

It feels like there ought to be an apostrophe to make FIRST possessive here, but that feels a bit odd. Instead, I'd propose this change, which also reflects the sentence structure of the other entries.

WPI – 1 individual representing the interests of Worcester Polytechnic University (WPI) which drove initial library development, continues to be reflected in the library name, and supplies a number of resources to support project activities.

WPILib Developers – 1 individual representing the community of WPILib developers.

> The FIRST and WPI representatives will be selected by their respective organizations and will have no defined term limits. The WPILib Developer member will be selected via nomination and vote of the WPILib Core Developers prior to the FIRST Championship each year and will serve a one year term running through the end of the next FIRST Championship. If more than 3 candidates are nominated, a run-off of the top 3 candidates from the initial vote will be held to determine the representative. If a WPILib Developer representative voluntarily resigns prior to the end of the term, the same process is followed to select a representative for the remainder of the term. There is no limit to the number of consecutive terms a WPILib Developer representative may serve.

Reviewer 3: A little morbid, Codex

### **Current Membership**

The current membership of the Steering Council is as follows:

FIRST Representative – Kevin O’Connor
WPI Representative – Amanda Bessette

> WPILib Developer Rep – Peter Johnson

Reviewer 4: ```suggestion
- FIRST Representative – Kevin O’Connor
- WPI Representative – Amanda Bessette
- WPILib Developer Rep – Peter Johnson
```
Fixes broken line breaks in the render by making it an unordered point list.

Core Developers meet \~weekly to discuss project status, current blocking items, etc.

## **WPILib Slack**

The WPILib Slack workspace is the primary internal coordination resource for WPILib development. This is a closed Slack Workspace whose membership consists of the Core Developers identified above along with members from other ecosystem partners who need to closely coordinate with WPILib (e.g. companies and third parties producing vendor libraries for use with WPILib). While this is a closed workspace, members will not generally be removed simply for becoming inactive as Core Developers. Members may be removed after becoming inactive in the workspace for greater than 1 year, or “for cause” if their continued participation in the workspace is deemed to be disruptive or detrimental to WPILib by the Steering Council.

## **Community Interaction**

WPILib Core Developers are allowed and encouraged to interact with the broader community of users and developers through any number of unofficial avenues including:

\-            In person interaction with teams via mentoring and/or event volunteering
\-            Forums such as the FTC Forum and Chief Delphi

> \-            Chat platforms such as various Slack and Discord spaces

Reviewer 4: ```suggestion
- In person interaction with teams via mentoring and/or event volunteering
- Forums such as the FTC Forum and Chief Delphi
- Chat platforms such as various Slack and Discord spaces
```
Fixes the broken line breaks by making it a unordered list.

## **Development Process**

This section aims to provide general guidelines on the development process for WPILib and its various projects without introducing any unnecessary bureaucratic burden on processes that have generally been working well for many years without formal guidelines.

While there is no clear hard delineation between these categories, generally:

\-            Bug Fixes: fix unintended behavior of the library and cause little to no breakage to user code.
\-            Minor Changes: add API methods or classes that complement existing methods and classes in the same project and do not cause significant breakage to user code or change to user experience.
\-            Significant Changes: either affect a large number of files or lines of code in an internal way, or may cause significant breaking changes to user code scoped to one small area of the library consisting of several classes or less, or narrowly scoped changes to user experience (e.g. reorganize one subsection of documentation, or change one set of buttons or workflow in a particular application).

> \-            Major changes: include things such as substantial new features, large re-writes of existing functionality, addition or removal of tools, or other changes with large implications on user code or experience.

Reviewer 4: ```suggestion
- Bug Fixes: fix unintended behavior of the library and cause little to no breakage to user code.
- Minor Changes: add API methods or classes that complement existing methods and classes in the same project and do not cause significant breakage to user code or change to user experience.
- Significant Changes: either affect a large number of files or lines of code in an internal way, or may cause significant breaking changes to user code scoped to one small area of the library consisting of several classes or less, or narrowly scoped changes to user experience (e.g. reorganize one subsection of documentation, or change one set of buttons or workflow in a particular application).
- Major changes: include things such as substantial new features, large re-writes of existing functionality, addition or removal of tools, or other changes with large implications on user code or experience.
```

### **Bug Fixes**

Bug fixes may be made by any developer, approved by any Core Developer with appropriate permissions on the associated project, and have no proposed minimum “dwell time” for feedback.

### **Minor Changes**

Minor changes may be made by any developer, approved by any Core Developer with appropriate permissions on the associated project and generally should remain open as a PR for a minimum of 24 hours to allow for any interested individuals to object with any concerns.

### **Significant Changes**

Significant changes proposed by non-Core Developers are recommended to solicit feedback via Issue or Discussion prior to performing any significant amount of work in order to ensure the proposed changes are aligned with the desired project direction. Significant changes, regardless of source, are encouraged to be submitted as Draft PRs in order to solicit feedback as early in the process as possible. Significant changes are encouraged to remain open for a minimum of:

\-            24 hours after being marked as ready if feedback has been provided on the Draft and appears to be in consensus

> \-            72 hours if no feedback has yet been provided or if the feedback does not appear to be in consensus.

Reviewer 4: ```suggestion
- 24 hours after being marked as ready if feedback has been provided on the Draft and appears to be in consensus
- 72 hours if no feedback has yet been provided or if the feedback does not appear to be in consensus.
```

### **Major Changes**

Major changes should generally be developed using the following process:

\-            Proposed via GitHub Discussion, Core Team discussion on Slack or during meetings, or Steering Council discussion during meetings
\-            Tracked via [Github Project board](https://github.com/orgs/wpilibsuite/projects)

> \-            Documented as a design doc in the [allwpilib repo](https://github.com/wpilibsuite/allwpilib/tree/main/design-docs) or other appropriate location for development and feedback. This should also include the proposed transition plan (timeline, deprecation vs hard-switch, etc.)

Reviewer 4: ```suggestion
1.  Proposed via GitHub Discussion, Core Team discussion on Slack or during meetings, or Steering Council discussion during meetings
2. Tracked via [Github Project board](https://github.com/orgs/wpilibsuite/projects)
3. Documented as a design doc in the [allwpilib repo](https://github.com/wpilibsuite/allwpilib/tree/main/design-docs) or other appropriate location for development and feedback. This should also include the proposed transition plan (timeline, deprecation vs hard-switch, etc.)
```
Makes this an ordered list, in addition to fixing broken line breaks.

Author: Thanks for the feedback. Silly me thinking the MD exported by Google Docs would be reasonable. I'll address these hopefully later today.

Reviewer: I think it was _technically_ reasonable when the line breaks were preserved by the trailing whitespace. Markdown is definitely a weird markup language.
